File: packages/runar-go/sdk_types.go
package runar

// ---------------------------------------------------------------------------
// SDK types for deploying and interacting with compiled Runar contracts on BSV
// ---------------------------------------------------------------------------

// UTXO represents an unspent transaction output.
type UTXO struct {
	Txid        string `json:"txid"`
	OutputIndex int    `json:"outputIndex"`
	Satoshis    int64  `json:"satoshis"`
	Script      string `json:"script"` // hex-encoded locking script
}

// TransactionData represents a parsed Bitcoin transaction (data shape for getTransaction return).
type TransactionData struct {
	Txid     string     `json:"txid"`
	Version  int        `json:"version"`
	Inputs   []TxInput  `json:"inputs"`
	Outputs  []TxOutput `json:"outputs"`
	Locktime int        `json:"locktime"`
	Raw      string     `json:"raw,omitempty"`
}

// TxInput represents a transaction input.
type TxInput struct {
	Txid        string `json:"txid"`
	OutputIndex int    `json:"outputIndex"`
	Script      string `json:"script"`   // hex-encoded scriptSig
	Sequence    uint32 `json:"sequence"`
}

// TxOutput represents a transaction output.
type TxOutput struct {
	Satoshis int64  `json:"satoshis"`
	Script   string `json:"script"` // hex-encoded locking script
}

// DeployOptions specifies options for deploying a contract.
type DeployOptions struct {
	Satoshis      int64  `json:"satoshis"`
	ChangeAddress string `json:"changeAddress,omitempty"`
}

// CallOptions specifies options for calling a contract method.
type CallOptions struct {
	Satoshis      int64                  `json:"satoshis,omitempty"`
	ChangeAddress string                 `json:"changeAddress,omitempty"`
	ChangePubKey  string                 `json:"changePubKey,omitempty"` // Override public key for change output (hex-encoded). Defaults to signer's pubkey.
	NewState      map[string]interface{} `json:"newState,omitempty"`

	// Multiple continuation outputs for multi-output methods (e.g., transfer).
	// Each entry specifies the satoshis and state for one output UTXO.
	// When provided, replaces the single continuation output from NewState.
	Outputs []OutputSpec `json:"outputs,omitempty"`

	// Additional contract UTXOs to include as inputs (e.g., for merge, swap,
	// or any multi-input spending pattern). Each UTXO's unlocking script uses
	// the same method and args as the primary call, with OP_PUSH_TX and Sig
	// auto-computed per input.
	AdditionalContractInputs []*UTXO `json:"additionalContractInputs,omitempty"`

	// Per-input args for additional contract inputs. When provided,
	// AdditionalContractInputArgs[i] overrides args for AdditionalContractInputs[i].
	// Sig params (nil) are still auto-computed per input.
	AdditionalContractInputArgs [][]interface{} `json:"additionalContractInputArgs,omitempty"`

	// Terminal outputs for methods that verify exact output structure via
	// extractOutputHash(). When set, the transaction is built with ONLY
	// the contract UTXO as input (no funding inputs, no change output).
	// The fee comes from the contract balance. The contract is considered
	// fully spent after this call (currentUtxo becomes nil).
	TerminalOutputs []TerminalOutput `json:"terminalOutputs,omitempty"`
}

// TerminalOutput specifies an exact output for a terminal method call.
type TerminalOutput struct {
	ScriptHex string `json:"scriptHex"`
	Satoshis  int64  `json:"satoshis"`
}

// OutputSpec specifies a single continuation output for multi-output calls.
type OutputSpec struct {
	Satoshis int64                  `json:"satoshis"`
	State    map[string]interface{} `json:"state"`
}

// PreparedCall holds all data from a prepared (but not yet signed) method call.
// Public fields are for external signer coordination. Internal fields (lowercase)
// are consumed by FinalizeCall().
type PreparedCall struct {
	// Public: callers use these to coordinate external signing
	Sighash     string `json:"sighash"`     // 64-char hex — BIP-143 hash external signers sign
	Preimage    string `json:"preimage"`    // hex — full BIP-143 preimage
	OpPushTxSig string `json:"opPushTxSig"` // hex — OP_PUSH_TX DER sig (empty if not needed)
	TxHex       string `json:"txHex"`       // hex — built TX (for backward compat / JSON serialization)
	SigIndices  []int  `json:"sigIndices"`  // which user-visible arg positions need external Sig values

	// Internal — consumed by FinalizeCall()
	methodName        string
	resolvedArgs      []interface{}
	methodSelectorHex string
	isStateful        bool
	isTerminal        bool
	needsOpPushTx     bool
	methodNeedsChange bool
	changePKHHex      string
	changeAmount      int64
	methodNeedsNewAmount bool
	newAmount         int64
	preimageIndex     int
	contractUtxo      UTXO
	newLockingScript  string
	newSatoshis       int64
	hasMultiOutput    bool
	contractOutputs   []ContractOutput
	codeSepIdx        int // adjusted OP_CODESEPARATOR byte offset, -1 if none
}

// ---------------------------------------------------------------------------
// Artifact types (compiled contract output)
// ---------------------------------------------------------------------------

// RunarArtifact is the compiled output of a Runar compiler.
type RunarArtifact struct {
	Version                string            `json:"version"`
	CompilerVersion        string            `json:"compilerVersion"`
	ContractName           string            `json:"contractName"`
	ABI                    ABI               `json:"abi"`
	Script                 string            `json:"script"`
	ASM                    string            `json:"asm"`
	StateFields            []StateField      `json:"stateFields,omitempty"`
	ConstructorSlots       []ConstructorSlot `json:"constructorSlots,omitempty"`
	BuildTimestamp         string            `json:"buildTimestamp"`
	CodeSeparatorIndex     *int              `json:"codeSeparatorIndex,omitempty"`
	CodeSeparatorIndices   []int             `json:"codeSeparatorIndices,omitempty"`
	ANF                    *ANFProgram       `json:"anf,omitempty"`
}

// ABI describes the contract's public interface.
type ABI struct {
	Constructor ABIConstructor `json:"constructor"`
	Methods     []ABIMethod    `json:"methods"`
}

// ABIConstructor describes the constructor parameters.
type ABIConstructor struct {
	Params []ABIParam `json:"params"`
}

// ABIMethod describes a contract method.
type ABIMethod struct {
	Name       string     `json:"name"`
	Params     []ABIParam `json:"params"`
	IsPublic   bool       `json:"isPublic"`
	IsTerminal *bool      `json:"isTerminal,omitempty"`
}

// ABIParam describes a single parameter.
type ABIParam struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

// StateField describes a state field in a stateful contract.
type StateField struct {
	Name         string      `json:"name"`
	Type         string      `json:"type"`
	Index        int         `json:"index"`
	InitialValue interface{} `json:"initialValue,omitempty"`
}

// ConstructorSlot describes where a constructor parameter placeholder
// resides in the compiled script (byte offset of the OP_0 placeholder).
type ConstructorSlot struct {
	ParamIndex int `json:"paramIndex"`
	ByteOffset int `json:"byteOffset"`
}

File: packages/runar-zig/src/sdk_types.zig
const std = @import("std");

// ---------------------------------------------------------------------------
// SDK types for deploying and interacting with compiled Runar contracts on BSV
// ---------------------------------------------------------------------------

/// UTXO represents an unspent transaction output.
pub const UTXO = struct {
    txid: []const u8, // 64-char hex txid
    output_index: i32,
    satoshis: i64,
    script: []const u8, // hex-encoded locking script

    pub fn clone(self: UTXO, allocator: std.mem.Allocator) !UTXO {
        return .{
            .txid = try allocator.dupe(u8, self.txid),
            .output_index = self.output_index,
            .satoshis = self.satoshis,
            .script = try allocator.dupe(u8, self.script),
        };
    }

    pub fn deinit(self: *UTXO, allocator: std.mem.Allocator) void {
        allocator.free(self.txid);
        allocator.free(self.script);
        self.* = .{ .txid = &.{}, .output_index = 0, .satoshis = 0, .script = &.{} };
    }
};

/// TransactionData represents a parsed Bitcoin transaction.
pub const TransactionData = struct {
    txid: []const u8,
    version: i32 = 1,
    inputs: []TxInput = &.{},
    outputs: []TxOutput = &.{},
    locktime: u32 = 0,
    raw: []const u8 = &.{},

    pub fn deinit(self: *TransactionData, allocator: std.mem.Allocator) void {
        if (self.txid.len > 0) allocator.free(self.txid);
        for (self.inputs) |*inp| {
            _ = inp;
        }
        if (self.inputs.len > 0) allocator.free(self.inputs);
        for (self.outputs) |*outp| {
            _ = outp;
        }
        if (self.outputs.len > 0) allocator.free(self.outputs);
        if (self.raw.len > 0) allocator.free(self.raw);
        self.* = .{ .txid = &.{} };
    }
};

/// TxInput represents a transaction input.
pub const TxInput = struct {
    txid: []const u8 = &.{},
    output_index: i32 = 0,
    script: []const u8 = &.{},
    sequence: u32 = 0xffffffff,
};

/// TxOutput represents a transaction output.
pub const TxOutput = struct {
    satoshis: i64 = 0,
    script: []const u8 = &.{}, // hex-encoded locking script
};

/// DeployOptions specifies options for deploying a contract.
pub const DeployOptions = struct {
    satoshis: i64,
    change_address: ?[]const u8 = null,
};

/// CallOptions specifies options for calling a contract method.
pub const CallOptions = struct {
    satoshis: i64 = 0,
    change_address: ?[]const u8 = null,
    new_state: ?[]const StateValue = null,
};

/// ContractOutput describes one contract continuation output.
pub const ContractOutput = struct {
    script: []const u8,
    satoshis: i64,
};

// ---------------------------------------------------------------------------
// Artifact types (compiled contract output)
// ---------------------------------------------------------------------------

/// RunarArtifact is the compiled output of a Runar compiler.
pub const RunarArtifact = struct {
    allocator: std.mem.Allocator,
    version: []const u8 = &.{},
    compiler_version: []const u8 = &.{},
    contract_name: []const u8 = &.{},
    abi: ABI = .{},
    script: []const u8 = &.{},
    asm_text: []const u8 = &.{},
    state_fields: []StateField = &.{},
    constructor_slots: []ConstructorSlot = &.{},
    code_sep_index_slots: []CodeSepIndexSlot = &.{},
    build_timestamp: []const u8 = &.{},
    code_separator_index: ?i32 = null,
    code_separator_indices: []i32 = &.{},
    anf_json: ?[]const u8 = null, // raw JSON of the ANF IR (for SDK auto-state computation)

    pub fn isStateful(self: *const RunarArtifact) bool {
        return self.state_fields.len > 0;
    }

    pub fn deinit(self: *RunarArtifact) void {
        const a = self.allocator;
        if (self.version.len > 0) a.free(self.version);
        if (self.compiler_version.len > 0) a.free(self.compiler_version);
        if (self.contract_name.len > 0) a.free(self.contract_name);
        if (self.script.len > 0) a.free(self.script);
        if (self.asm_text.len > 0) a.free(self.asm_text);
        if (self.build_timestamp.len > 0) a.free(self.build_timestamp);
        self.abi.deinit(a);
        for (self.state_fields) |*sf| sf.deinit(a);
        if (self.state_fields.len > 0) a.free(self.state_fields);
        if (self.constructor_slots.len > 0) a.free(self.constructor_slots);
        if (self.code_sep_index_slots.len > 0) a.free(self.code_sep_index_slots);
        if (self.code_separator_indices.len > 0) a.free(self.code_separator_indices);
        if (self.anf_json) |aj| a.free(aj);
        self.* = .{ .allocator = a };
    }

    /// Parse a RunarArtifact from JSON text.
    pub fn fromJson(allocator: std.mem.Allocator, json_text: []const u8) !RunarArtifact {
        var artifact = RunarArtifact{ .allocator = allocator };
        errdefer artifact.deinit();

        var parsed = try std.json.parseFromSlice(std.json.Value, allocator, json_text, .{});
        defer parsed.deinit();

        const root = parsed.value.object;

        if (root.get("contractName")) |v| {
            if (v == .string) artifact.contract_name = try allocator.dupe(u8, v.string);
        }
        if (root.get("version")) |v| {
            if (v == .string) artifact.version = try allocator.dupe(u8, v.string);
        }
        if (root.get("compilerVersion")) |v| {
            if (v == .string) artifact.compiler_version = try allocator.dupe(u8, v.string);
        }
        if (root.get("script")) |v| {
            if (v == .string) artifact.script = try allocator.dupe(u8, v.string);
        }
        if (root.get("asm")) |v| {
            if (v == .string) artifact.asm_text = try allocator.dupe(u8, v.string);
        }
        if (root.get("buildTimestamp")) |v| {
            if (v == .string) artifact.build_timestamp = try allocator.dupe(u8, v.string);
        }
        if (root.get("codeSeparatorIndex")) |v| {
            if (v == .integer) artifact.code_separator_index = @intCast(v.integer);
        }
        if (root.get("codeSeparatorIndices")) |v| {
            if (v == .array) {
                const items = v.array.items;
                var indices = try allocator.alloc(i32, items.len);
                for (items, 0..) |item, i| {
                    indices[i] = if (item == .integer) @intCast(item.integer) else 0;
                }
                artifact.code_separator_indices = indices;
            }
        }

        // Parse ABI
        if (root.get("abi")) |abi_val| {
            if (abi_val == .object) {
                artifact.abi = try ABI.fromJsonValue(allocator, abi_val.object);
            }
        }

        // Parse stateFields
        if (root.get("stateFields")) |sf_val| {
            if (sf_val == .array) {
                const items = sf_val.array.items;
                var fields = try allocator.alloc(StateField, items.len);
                for (items, 0..) |item, i| {
                    fields[i] = try StateField.fromJsonValue(allocator, item.object);
                }
                artifact.state_fields = fields;
            }
        }

        // Store raw ANF JSON for the SDK ANF interpreter
        if (root.get("anf")) |anf_val| {
            if (anf_val == .object) {
                // Re-stringify the ANF object so the interpreter can parse it independently
                const anf_str = std.json.Stringify.valueAlloc(allocator, anf_val, .{}) catch null;
                if (anf_str) |s| {
                    artifact.anf_json = s;
                }
            }
        }

        // Parse constructorSlots
        if (root.get("constructorSlots")) |cs_val| {
            if (cs_val == .array) {
                const items = cs_val.array.items;
                var slots = try allocator.alloc(ConstructorSlot, items.len);
                for (items, 0..) |item, i| {
                    const obj = item.object;
                    slots[i] = .{
                        .param_index = if (obj.get("paramIndex")) |pi| @intCast(pi.integer) else 0,
                        .byte_offset = if (obj.get("byteOffset")) |bo| @intCast(bo.integer) else 0,
                    };
                }
                artifact.constructor_slots = slots;
            }
        }

        // Parse codeSepIndexSlots
        if (root.get("codeSepIndexSlots")) |csis_val| {
            if (csis_val == .array) {
                const items = csis_val.array.items;
                var csis = try allocator.alloc(CodeSepIndexSlot, items.len);
                for (items, 0..) |item, i| {
                    const obj = item.object;
                    csis[i] = .{
                        .byte_offset = if (obj.get("byteOffset")) |bo| @intCast(bo.integer) else 0,
                        .code_sep_index = if (obj.get("codeSepIndex")) |ci| @intCast(ci.integer) else 0,
                    };
                }
                artifact.code_sep_index_slots = csis;
            }
        }

        return artifact;
    }
};

/// ABI describes the contract's public interface.
pub const ABI = struct {
    constructor: ABIConstructor = .{},
    methods: []ABIMethod = &.{},

    pub fn deinit(self: *ABI, allocator: std.mem.Allocator) void {
        self.constructor.deinit(allocator);
        for (self.methods) |*m| m.deinit(allocator);
        if (self.methods.len > 0) allocator.free(self.methods);
        self.* = .{};
    }

    pub fn fromJsonValue(allocator: std.mem.Allocator, obj: std.json.ObjectMap) !ABI {
        var abi = ABI{};
        errdefer abi.deinit(allocator);

        if (obj.get("constructor")) |ctor_val| {
            if (ctor_val == .object) {
                abi.constructor = try ABIConstructor.fromJsonValue(allocator, ctor_val.object);
            }
        }

        if (obj.get("methods")) |methods_val| {
            if (methods_val == .array) {
                const items = methods_val.array.items;
                var methods = try allocator.alloc(ABIMethod, items.len);
                errdefer allocator.free(methods);
                for (items, 0..) |item, i| {
                    methods[i] = try ABIMethod.fromJsonValue(allocator, item.object);
                }
                abi.methods = methods;
            }
        }

        return abi;
    }
};

/// ABIConstructor describes the constructor parameters.
pub const ABIConstructor = struct {
    params: []ABIParam = &.{},

    pub fn deinit(self: *ABIConstructor, allocator: std.mem.Allocator) void {
        for (self.params) |*p| p.deinit(allocator);
        if (self.params.len > 0) allocator.free(self.params);
        self.* = .{};
    }

    pub fn fromJsonValue(allocator: std.mem.Allocator, obj: std.json.ObjectMap) !ABIConstructor {
        var ctor = ABIConstructor{};
        if (obj.get("params")) |params_val| {
            if (params_val == .array) {
                const items = params_val.array.items;
                var params = try allocator.alloc(ABIParam, items.len);
                for (items, 0..) |item, i| {
                    params[i] = try ABIParam.fromJsonValue(allocator, item.object);
                }
                ctor.params = params;
            }
        }
        return ctor;
    }
};

/// ABIMethod describes a contract method.
pub const ABIMethod = struct {
    name: []const u8 = &.{},
    params: []ABIParam = &.{},
    is_public: bool = false,
    is_terminal: ?bool = null,

    pub fn deinit(self: *ABIMethod, allocator: std.mem.Allocator) void {
        if (self.name.len > 0) allocator.free(self.name);
        for (self.params) |*p| p.deinit(allocator);
        if (self.params.len > 0) allocator.free(self.params);
        self.* = .{};
    }

    pub fn fromJsonValue(allocator: std.mem.Allocator, obj: std.json.ObjectMap) !ABIMethod {
        var method = ABIMethod{};
        errdefer method.deinit(allocator);

        if (obj.get("name")) |v| {
            if (v == .string) method.name = try allocator.dupe(u8, v.string);
        }
        if (obj.get("isPublic")) |v| {
            if (v == .bool) method.is_public = v.bool;
        }
        if (obj.get("isTerminal")) |v| {
            if (v == .bool) method.is_terminal = v.bool;
        }
        if (obj.get("params")) |params_val| {
            if (params_val == .array) {
                const items = params_val.array.items;
                var params = try allocator.alloc(ABIParam, items.len);
                for (items, 0..) |item, i| {
                    params[i] = try ABIParam.fromJsonValue(allocator, item.object);
                }
                method.params = params;
            }
        }

        return method;
    }
};

/// ABIParam describes a single parameter.
pub const ABIParam = struct {
    name: []const u8 = &.{},
    type_name: []const u8 = &.{},

    pub fn deinit(self: *ABIParam, allocator: std.mem.Allocator) void {
        if (self.name.len > 0) allocator.free(self.name);
        if (self.type_name.len > 0) allocator.free(self.type_name);
        self.* = .{};
    }

    pub fn fromJsonValue(allocator: std.mem.Allocator, obj: std.json.ObjectMap) !ABIParam {
        var param = ABIParam{};
        errdefer param.deinit(allocator);

        if (obj.get("name")) |v| {
            if (v == .string) param.name = try allocator.dupe(u8, v.string);
        }
        if (obj.get("type")) |v| {
            if (v == .string) param.type_name = try allocator.dupe(u8, v.string);
        }

        return param;
    }
};

/// StateField describes a state field in a stateful contract.
pub const StateField = struct {
    name: []const u8 = &.{},
    type_name: []const u8 = &.{},
    index: i32 = 0,
    initial_value: ?[]const u8 = null, // stored as string representation

    pub fn deinit(self: *StateField, allocator: std.mem.Allocator) void {
        if (self.name.len > 0) allocator.free(self.name);
        if (self.type_name.len > 0) allocator.free(self.type_name);
        if (self.initial_value) |iv| allocator.free(iv);
        self.* = .{};
    }

    pub fn fromJsonValue(allocator: std.mem.Allocator, obj: std.json.ObjectMap) !StateField {
        var field = StateField{};
        errdefer field.deinit(allocator);

        if (obj.get("name")) |v| {
            if (v == .string) field.name = try allocator.dupe(u8, v.string);
        }
        if (obj.get("type")) |v| {
            if (v == .string) field.type_name = try allocator.dupe(u8, v.string);
        }
        if (obj.get("index")) |v| {
            if (v == .integer) field.index = @intCast(v.integer);
        }
        if (obj.get("initialValue")) |v| {
            switch (v) {
                .string => |s| field.initial_value = try allocator.dupe(u8, s),
                .integer => |n| {
                    field.initial_value = try std.fmt.allocPrint(allocator, "{d}", .{n});
                },
                .bool => |b| {
                    field.initial_value = try allocator.dupe(u8, if (b) "true" else "false");
                },
                else => {},
            }
        }

        return field;
    }
};

/// ConstructorSlot describes where a constructor parameter placeholder
/// resides in the compiled script (byte offset of the OP_0 placeholder).
pub const ConstructorSlot = struct {
    param_index: i32,
    byte_offset: i32,
};

/// CodeSepIndexSlot describes where a codeSeparatorIndex placeholder (OP_0)
/// resides in the template script. The SDK substitutes these at deployment
/// time with the adjusted codeSeparatorIndex value that accounts for
/// constructor arg expansion.
pub const CodeSepIndexSlot = struct {
    byte_offset: i32,
    code_sep_index: i32,
};

// ---------------------------------------------------------------------------
// StateValue — dynamically-typed value used in contract state
// ---------------------------------------------------------------------------

/// StateValue is a tagged union for contract state values and method arguments.
pub const StateValue = union(enum) {
    int: i64,
    boolean: bool,
    bytes: []const u8, // hex-encoded

    pub fn deinit(self: StateValue, allocator: std.mem.Allocator) void {
        switch (self) {
            .bytes => |b| allocator.free(b),
            else => {},
        }
    }

    pub fn clone(self: StateValue, allocator: std.mem.Allocator) !StateValue {
        return switch (self) {
            .int => |n| .{ .int = n },
            .boolean => |b| .{ .boolean = b },
            .bytes => |b| .{ .bytes = try allocator.dupe(u8, b) },
        };
    }
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "RunarArtifact.fromJson parses minimal artifact" {
    const json =
        \\{"contractName":"Test","version":"1","compilerVersion":"1.0","script":"5100","asm":"OP_1 OP_0",
        \\"abi":{"constructor":{"params":[]},"methods":[{"name":"unlock","params":[],"isPublic":true}]},
        \\"stateFields":[],"constructorSlots":[],"buildTimestamp":"2024-01-01"}
    ;
    var artifact = try RunarArtifact.fromJson(std.testing.allocator, json);
    defer artifact.deinit();

    try std.testing.expectEqualStrings("Test", artifact.contract_name);
    try std.testing.expectEqualStrings("5100", artifact.script);
    try std.testing.expectEqual(@as(usize, 1), artifact.abi.methods.len);
    try std.testing.expectEqualStrings("unlock", artifact.abi.methods[0].name);
    try std.testing.expect(artifact.abi.methods[0].is_public);
}

test "RunarArtifact.fromJson parses stateful artifact with constructor slots" {
    const json =
        \\{"contractName":"Counter","version":"1","compilerVersion":"1.0","script":"005100","asm":"OP_0 OP_1 OP_0",
        \\"abi":{"constructor":{"params":[{"name":"count","type":"int"}]},"methods":[{"name":"increment","params":[],"isPublic":true}]},
        \\"stateFields":[{"name":"count","type":"int","index":0}],
        \\"constructorSlots":[{"paramIndex":0,"byteOffset":0}],
        \\"codeSeparatorIndex":2,"buildTimestamp":"2024-01-01"}
    ;
    var artifact = try RunarArtifact.fromJson(std.testing.allocator, json);
    defer artifact.deinit();

    try std.testing.expectEqualStrings("Counter", artifact.contract_name);
    try std.testing.expectEqual(@as(usize, 1), artifact.state_fields.len);
    try std.testing.expectEqualStrings("count", artifact.state_fields[0].name);
    try std.testing.expectEqual(@as(usize, 1), artifact.constructor_slots.len);
    try std.testing.expectEqual(@as(i32, 0), artifact.constructor_slots[0].byte_offset);
    try std.testing.expectEqual(@as(i32, 2), artifact.code_separator_index.?);
    try std.testing.expect(artifact.isStateful());
}
